### PR TITLE
[JUP-120] Update logging levels

### PR DESCRIPTION
**Pull request by tommycodebox**
## 🧠  Description
- Use simple logger for all logging levels

image

### CODE REVIEW

File: index.js
Comment (arijoon) [inline index.js:67]: could an `Error` object have more properties than these 3 that were added as context?

Comment (arijoon) [reply index.js:67]: Actually ignore me there, it seems that is not even a valid use case of `Error`, and we have some of that in our codebase. An extension of `Error` could have more properties however so may want to add a `context` field there that we can populate in our custom errors such as `DVFError`

File: package.json
Comment (arijoon) [inline package.json:10]: can remove this dep if no longer used

File: index.js
Comment (arijoon) [inline index.js:102]: do we no longer need to print anything to stderr?

Comment (tommycodebox) [reply index.js:102]: I dont think we need, if we dump json with correct fields and GCP parses that

Comment (tommycodebox) [reply index.js:102]: and I checked codebase where this package was required and there were no cases when options were passed

**Review by arijoon — APPROVED** [review]

File: index.js
Comment (adrian-gierakowski) [reply index.js:70]: There are many error classes in libraries that include extra props (like MongoError) which might be useful to long.

Comment (adrian-gierakowski) [reply index.js:70]: Btw. I'm not sure about simply concatenating all args into a string is sticking it all into message. I think we should be able to pass in objects and have those logged in structured way.

File: index.js
Comment (adrian-gierakowski) [inline index.js:81]: I'd like to be able to toggle (with an env var) between logging as single line of JSON and pretty printing. This is really useful when running things locally.

File: package.json
Comment (tommycodebox) [reply package.json:10]: I removed that `.enabled` at all, as we dont seem to use it, and I created simple pattern to account for `DEBUG` env var to allow logging all levels with `DEBUG=*` and specific levels in isolation e.g. `DEBUG=WARN`

by default we would log following levels (basically all without DEBUG level)
`[ LOG, WARN, ERROR, EMERGENCY ]`

File: index.js
Comment (tommycodebox) [reply index.js]: so now by adding `PRETTY=true` you will get json indented with 2 places

File: index.js
Comment (arijoon) [inline index.js:96]: Can we use a different logger format in case of `PRETTY`? 
```
timestamp | context: message | JSON.stringify(args)
```
instead of a json dump of everything to make it much more readable locally since it won't be auto parsed

File: package.json
Comment (adrian-gierakowski) [reply package.json:10]: I'm sorry, I'm not going to approve this PR if we don't get feature parity with existing logger. It's very simple the achieve this: bring back the debug module and use it's `.enabled` prop to check if you should log or not (even though you don's use it for actual logging).

File: index.js
Comment (adrian-gierakowski) [inline index.js:57]: after these changes, the lazy loggers don't do what they were designed to do: `expandThunks` should have only been called if logger was enabled

**Review by adrian-gierakowski — CHANGES_REQUESTED** [review]
the following comment has not been sufficiently addressed:
https://github.com/DeversiFi/logger/pull/2/files#r734354894

File: index.js
Comment (adrian-gierakowski) [inline index.js:95]: llthis is not efficient: this loggic is executed every single time the logging function is called, even though eventually we might end up not logging anything. Original logger implementation decides if particular logger instance is `enabled` or not at the instantiation time (once per process run) and then simply checks `.enabled` prop to decide if logging logic should be executed or not

File: package.json
Comment (adrian-gierakowski) [reply package.json:10]: btw. if we use `debug` module (and/or `DEBUG` env var) we'll need the `dvf:` prefix as there are many other modules in the JS ecosystem using `debug` and if we don't prefix we might accidentally enable logging in one of the libraries we used. Keeping it backward compatible will also save you time since you will not need to update env vars in existing services which use `dvf-logger`

**Comment by tommycodebox** [comment]
@adrian-gierakowski I have addressed your feedback by optimizing the logger using `.enabled` but without using `debug`, logic now also wont log in test environment

**Comment by arijoon** [comment]
> @adrian-gierakowski I have addressed your feedback by optimizing the logger using `.enabled` but without using `debug`, logic now also wont log in test environment

I don't think that will bring back the feature @adrian-gierakowski was asking for which is pattern matching the files to log based on their name, e.g. `DEBUG='dvf:ERROR:*:serviceX.js'`

**Comment by tommycodebox** [comment]
ok, if we are using this feature, lets then use `.enabled` of that `debug` package to determine whether to log or not 👍🏻

**Comment by arijoon** [comment]
Flattening of the arguments object: https://darcs.slack.com/archives/GL7JLS5RD/p1637682852493000

**Comment by arijoon** [comment]
Is this ready for merge?

**Comment by tommycodebox** [comment]
@arijoon this is finished from my side but was still pending review from you and @adrian-gierakowski so we are clear on this

File: index.js
Comment (arijoon) [inline index.js:57]: does this need to be defaulted?

File: stringify.js
Comment (arijoon) [inline stringify.js:3]: `'undefined'` would be more appropriate

**Review by arijoon — APPROVED** [review]
Looks great to me @tommycodebox , lets put it in stark trading `dev` and see how it looks in action

File: yarn.lock
Comment (arijoon) [inline yarn.lock:17]: btw this doesn't seem to be used anymore so can it be removed?

File: stringify.js
Comment (tommycodebox) [reply stringify.js:3]: JSON doesnt support `undefined`, but does support `null` and Im not sure we need a string of `'undefined'` 🤔

File: index.js
Comment (tommycodebox) [reply index.js:57]: nope, its defaulted in `stringify` function

**Pull request merged**
